src/parser/statement.js

```diff
@@ -1009,12 +1019,36 @@ export default class StatementParser extends ExpressionParser {
       return;
     }
 
+    const isPrivate = this.eat(tt.hash);
+
+    if (isPrivate) {
+      this.expectOnePlugin(["classPrivateProperties", "classPrivateMethods"]);
+    }
+
     const isSimple = this.match(tt.name);
-    const key = this.parseClassPropertyName(methodOrProp);
+    const key = isPrivate
+      ? this.parseClassPrivateName(methodOrProp)
+      : this.parseClassPropertyName(methodOrProp);
 
     this.parsePostMemberNameModifiers(methodOrProp);
 
     if (this.isClassMethod()) {
+      if (isPrivate) {
+        // TODO: seems a bit inconsistent error throwing: in some cases we've already
+        // eaten the #, but in other cases we haven't. How to solve?
+        this.expectPlugin("classPrivateMethods");
+        // private "normal" method
+        method.kind = "method";
```

Comment:
Can we hoist this to the beginning of the `this.isClassMethod()` if statement?

Reply:
Confused about where exactly you want this part to be moved to.

Reply:
Before the `if (isPrivate)`.